File: frontend/mock.js
/* =============================================================================
 * mock.js —— B 线的契约一致假数据源 + 契约校验器（纯逻辑，无 DOM）
 * =============================================================================
 * 为什么要有这个文件（而不是全塞进 app.js）：
 *   1. 它是**跨语言契约检查点**：本文件产出的帧会被 backend/contract.py 校验
 *      （node frontend/mock.js --limit 6 | python tools/check_contract.py）。
 *      A 线的 Python、B 线的 JS 必须对同一份契约得出同样的结论，
 *      否则 M2 把真数据接进来时才发现字段不一致 —— 那时改就晚了。
 *   2. 浏览器直接双击 index.html 时（没有后端），前端靠它兜底演示六态。
 *
 * ⚠️ 本文件里 status→指标的映射必须与 backend/mock.py 保持同一套语义：
 *    状态写"疲劳"时数字就必须疲劳。改一边必须改另一边。
 * ========================================================================== */
(function (root) {
  "use strict";

  var STATUS_VALUES = ["normal", "fatigue_risk", "adjust_posture", "unreliable", "disconnected", "done"];
  var BLINK_STATES = ["OPEN", "CLOSING", "CLOSED", "OPENING"];
  var VITAL_KEYS = ["hr_bpm", "hr_conf", "rr_per_min", "rr_conf"];
  var TOP_KEYS = ["ts", "frame_id", "face", "behavior", "vital", "quality", "status", "advice", "reason"];
  var DEMO_SEQUENCE = STATUS_VALUES.slice();

  var STATUS_ZH = {
    normal: "正常",
    fatigue_risk: "疲劳风险升高",
    adjust_posture: "请调整姿势",
    unreliable: "信号不可靠",
    disconnected: "连接中断",
    done: "测量完成"
  };

  var ADVICE = {
    normal: ["状态正常", "各项指标均在正常范围，信号质量良好"],
    fatigue_risk: ["疲劳风险升高，建议休息 5 分钟或远眺放松", "触发：PERCLOS 偏高；长闭眼次数增加"],
    adjust_posture: ["请正对摄像头，保持面部完整出现在画面内", "人脸可见率或头部姿态超出允许范围"],
    unreliable: ["信号不可靠，暂不输出测量结果", "光照/运动导致信号质量低于门控阈值"],
    disconnected: ["连接中断，请检查视频源或后端服务", "超过 ws_disconnect_timeout_s 未收到数据帧"],
    done: ["测量完成", "本次测量正常结束，结果已归档"]
  };

  /* ---- 确定性 PRNG（mulberry32）：同 seed 必得同序列 ---- */
  function makeRng(seed) {
    var s = seed >>> 0;
    return function () {
      s = (s + 0x6d2b79f5) >>> 0;
      var t = s;
      t = Math.imul(t ^ (t >>> 15), t | 1);
      t ^= t + Math.imul(t ^ (t >>> 7), t | 61);
      return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
    };
  }

  function rngFor(seed, frameId) {
    // 与 backend/mock.py 的 _rng 同思路：每帧一个独立、可复现的随机源
    return makeRng((Math.imul(seed, 1000003) ^ Math.imul(frameId, 2654435761)) >>> 0);
  }

  function round(x, n) {
    var f = Math.pow(10, n);
    return Math.round(x * f) / f;
  }
  function clamp01(x) {
    return x < 0 ? 0 : x > 1 ? 1 : x;
  }

  /* --------------------------------------------------------------------------
   * mockFrame —— 按 status 反推指标值，保证"状态与数字自洽"
   * ------------------------------------------------------------------------ */
  function mockFrame(frameId, opts) {
    opts = opts || {};
    var seed = opts.seed === undefined ? 20260910 : opts.seed;
    var statuses = opts.statuses || DEMO_SEQUENCE;
    var status = opts.status || statuses[((frameId % statuses.length) + statuses.length) % statuses.length];
    if (STATUS_VALUES.indexOf(status) < 0) throw new Error("未知 status: " + status);

    var r = rngFor(seed, frameId);
    var adv = ADVICE[status];

    var visible = Math.min(0.999, 0.96 + (r() - 0.5) * 0.04);
    var bbox = [320 + Math.round((r() - 0.5) * 4), 180 + Math.round((r() - 0.5) * 4),
                180 + Math.round((r() - 0.5) * 6), 220 + Math.round((r() - 0.5) * 6)];
    var yaw = (r() - 0.5) * 6, pitch = (r() - 0.5) * 4, roll = (r() - 0.5) * 4;

    var earLeft = 0.27 + (r() - 0.5) * 0.04;
    var earRight = earLeft + (r() - 0.5) * 0.02;
    var blinkState = "OPEN";
    var blinkCount = 12 + Math.floor(frameId / 20);
    var blinkRate = 17 + (r() - 0.5) * 4;
    var perclos = 0.05 + r() * 0.03;
    var longClose = 0;
    var mar = 0.10 + r() * 0.05;
    var yawn = 0;
    var light = 0.82 + (r() - 0.5) * 0.10;
    var motion = 0.03 + r() * 0.03;
    var overall = 0.86 + (r() - 0.5) * 0.08;
    var vital = {};

    if (status === "fatigue_risk") {
      perclos = 0.30 + r() * 0.12;
      blinkRate = 9 + (r() - 0.5) * 4;
      longClose = 1 + Math.floor(r() * 3);
      yawn = 2 + Math.floor(r() * 3);
      earLeft = 0.20 + (r() - 0.5) * 0.05;
      earRight = earLeft + (r() - 0.5) * 0.02;
      blinkState = r() < 0.35 ? "CLOSED" : "OPEN";
    } else if (status === "adjust_posture") {
      visible = 0.35 + r() * 0.25;
      yaw = (32 + r() * 14) * (r() < 0.5 ? 1 : -1);
      pitch = (r() - 0.5) * 12;
    } else if (status === "unreliable") {
      light = 0.22 + r() * 0.22;
      motion = 0.45 + r() * 0.40;
      overall = 0.30 + r() * 0.25;
    } else if (status === "disconnected") {
      visible = 0; bbox = [0, 0, 0, 0];
      yaw = pitch = roll = 0;
      earLeft = earRight = 0;
      perclos = 0; blinkRate = 0; mar = 0;
      light = motion = overall = 0;
    } else if (status === "done") {
      perclos = 0.08; blinkRate = 16;
    }

    if ((status === "normal" || status === "fatigue_risk" || status === "done") && overall >= 0.75) {
      vital = {
        hr_bpm: round(68 + (r() - 0.5) * 10, 1),
        hr_conf: round(Math.min(0.98, 0.62 + overall * 0.35 + (r() - 0.5) * 0.1), 3),
        rr_per_min: round(15 + (r() - 0.5) * 4, 1),
        rr_conf: round(Math.min(0.95, 0.55 + overall * 0.3 + (r() - 0.5) * 0.1), 3)
      };
    }

    return {
      ts: opts.ts === undefined || opts.ts === null ? round(Date.now() / 1000, 3) : opts.ts,
      frame_id: frameId,
      face: { visible: round(visible, 3), bbox: bbox, pose: { yaw: round(yaw, 2), pitch: round(pitch, 2), roll: round(roll, 2) } },
      behavior: {
        ear_left: round(earLeft, 4), ear_right: round(earRight, 4), blink_state: blinkState,
        blink_count: blinkCount, blink_rate_per_min: round(blinkRate, 2), perclos: round(perclos, 4),
        long_close_count: longClose, mar: round(mar, 4), yawn_count: yawn
      },
      vital: {
        hr_bpm: vital.hr_bpm === undefined ? null : vital.hr_bpm,
        hr_conf: vital.hr_conf === undefined ? null : vital.hr_conf,
        rr_per_min: vital.rr_per_min === undefined ? null : vital.rr_per_min,
        rr_conf: vital.rr_conf === undefined ? null : vital.rr_conf
      },
      quality: { light_score: round(clamp01(light), 4), motion_score: round(clamp01(motion), 4), overall: round(clamp01(overall), 4) },
      status: status,
      advice: adv[0],
      reason: adv[1]
    };
  }

  /* --------------------------------------------------------------------------
   * validateFrame —— 与 backend/contract.py 同规则的校验器
   * 目的：A 线的 Python 与 B 线的 JS 对契约的判断必须一致。
   * ------------------------------------------------------------------------ */
  function sameKeys(obj, keys) {
    // 集合比较，不要写死排序后的字符串 —— 写死顺序时 rr_conf/rr_per_min 的字典序
    // 与声明序不同，会制造"合法的帧被判非法"这种最难查的假阳性（这里踩过）。
    var a = Object.keys(obj).sort();
    var b = keys.slice().sort();
    return a.length === b.length && a.every(function (k, i) { return k === b[i]; });
  }

  function validateFrame(f) {
    var e = [];
    if (typeof f !== "object" || f === null || Array.isArray(f)) return ["顶层必须是 object"];

    TOP_KEYS.forEach(function (k) { if (!(k in f)) e.push("缺少顶层字段：" + k); });
    Object.keys(f).forEach(function (k) { if (TOP_KEYS.indexOf(k) < 0) e.push("出现契约外顶层字段：" + k); });

    if ("ts" in f && typeof f.ts !== "number") e.push("ts 必须是数字");
    if ("frame_id" in f && !Number.isInteger(f.frame_id)) e.push("frame_id 必须是整数");

    var face = f.face;
    if (typeof face !== "object" || face === null) e.push("face 必须是 object");
    else {
      if (!sameKeys(face, ["visible", "bbox", "pose"])) {
        e.push("face 字段应恰为 visible/bbox/pose，实际 " + Object.keys(face).sort().join("/"));
      }
      if (typeof face.visible !== "number" || face.visible < 0 || face.visible > 1) e.push("face.visible 必须是 0~1 的数");
      if (!Array.isArray(face.bbox) || face.bbox.length !== 4 || !face.bbox.every(Number.isInteger)) {
        e.push("face.bbox 必须是 4 个整数的数组 [x, y, w, h]");
      }
      if (typeof face.pose !== "object" || face.pose === null) e.push("face.pose 必须是 object");
      else {
        if (!sameKeys(face.pose, ["yaw", "pitch", "roll"])) {
          e.push("face.pose 字段应恰为 yaw/pitch/roll，实际 " + Object.keys(face.pose).sort().join("/"));
        }
        ["yaw", "pitch", "roll"].forEach(function (k) {
          if (typeof face.pose[k] !== "number") e.push("face.pose." + k + " 必须是数字");
        });
      }
    }

    var NEED_BEH = ["ear_left", "ear_right", "blink_state", "blink_count", "blink_rate_per_min",
                    "perclos", "long_close_count", "mar", "yawn_count"];
    var beh = f.behavior;
    if (typeof beh !== "object" || beh === null) e.push("behavior 必须是 object");
    else {
      if (!sameKeys(beh, NEED_BEH)) {
        e.push("behavior 字段不符：实际 " + Object.keys(beh).sort().join("/"));
      }
      if (BLINK_STATES.indexOf(beh.blink_state) < 0) e.push("behavior.blink_state 必须是 " + BLINK_STATES.join("/"));
    }

    if (typeof f.vital !== "object" || f.vital === null) e.push("vital 必须是 object（没有数据用 null 填充，不要省略键）");
    else {
      if (!sameKeys(f.vital, VITAL_KEYS)) {
        e.push("vital 字段应恰为 " + VITAL_KEYS.join("/") + "，实际 " + Object.keys(f.vital).sort().join("/"));
      }
      ["hr_conf", "rr_conf"].forEach(function (k) {
        var v = f.vital[k];
        if (v !== null && (typeof v !== "number" || v < 0 || v > 1)) e.push("vital." + k + " 是置信度，必须在 0~1");
      });
    }

    var q = f.quality;
    if (typeof q !== "object" || q === null || !sameKeys(q, ["light_score", "motion_score", "overall"])) {
      e.push("quality 字段应恰为 light_score/motion_score/overall");
    } else {
      ["light_score", "motion_score", "overall"].forEach(function (k) {
        if (typeof q[k] !== "number" || q[k] < 0 || q[k] > 1) e.push("quality." + k + " 必须在 0~1");
      });
    }

    if (STATUS_VALUES.indexOf(f.status) < 0) e.push("status 必须是 " + STATUS_VALUES.join("/"));
    ["advice", "reason"].forEach(function (k) {
      if (typeof f[k] !== "string" || !f[k].trim()) e.push(k + " 必须是非空字符串");
    });

    return e;
  }

  /* 契约自检：六态各产一帧 + 故意造 5 个坏帧，校验器必须抓住 */
  function selfTest(seed) {
    var failures = [];
    var checked = 0;
    DEMO_SEQUENCE.forEach(function (st, i) {
      var fr = mockFrame(i, { status: st, seed: seed });
      var errs = validateFrame(fr);
      checked++;
      if (errs.length) failures.push("六态帧 " + st + " 不合法：" + errs.join("；"));
    });

    var base = mockFrame(0, { status: "normal", seed: seed });
    function clone(o) { return JSON.parse(JSON.stringify(o)); }
    var bad = [
      ["未知 status", (function () { var x = clone(base); x.status = "sleeping"; return x; })()],
      ["bbox 少一个", (function () { var x = clone(base); x.face.bbox = [320, 180, 180]; return x; })()],
      ["多出契约外字段", (function () { var x = clone(base); x.debug = 1; return x; })()],
      ["缺 quality", (function () { var x = clone(base); delete x.quality; return x; })()],
      ["visible 越界", (function () { var x = clone(base); x.face.visible = 1.4; return x; })()]
    ];
    bad.forEach(function (pair) {
      checked++;
      if (validateFrame(pair[1]).length === 0) failures.push("校验器漏检：" + pair[0]);
    });
    return { checked: checked, failures: failures, ok: failures.length === 0 };
  }

  function stream(hz, limit, seed, statuses) {
    var out = [];
    var period = hz > 0 ? 1 / hz : 0;
    var n = limit === undefined || limit === null ? 6 : limit;
    for (var i = 0; i < n; i++) {
      out.push(mockFrame(i, { seed: seed, statuses: statuses, ts: period ? round(i / hz, 3) : null }));
    }
    return out;
  }

  var API = {
    STATUS_VALUES: STATUS_VALUES,
    BLINK_STATES: BLINK_STATES,
    VITAL_KEYS: VITAL_KEYS,
    TOP_KEYS: TOP_KEYS,
    DEMO_SEQUENCE: DEMO_SEQUENCE,
    STATUS_ZH: STATUS_ZH,
    mockFrame: mockFrame,
    validateFrame: validateFrame,
    selfTest: selfTest,
    stream: stream
  };

  if (typeof module !== "undefined" && module.exports) module.exports = API;   // node
  root.VigiLensMock = API;                                                      // browser
})(typeof globalThis !== "undefined" ? globalThis : this);

/* ---- node CLI：给跨语言契约检查用 ----
 *   node frontend/mock.js --limit 6            → 每行一帧 JSON
 *   node frontend/mock.js --selftest           → 跑 JS 侧契约自检
 */
if (typeof require !== "undefined" && typeof module !== "undefined" && require.main === module) {
  var M = module.exports;
  var argv = process.argv.slice(2);
  function arg(name, dflt) {
    var i = argv.indexOf(name);
    return i >= 0 && i + 1 < argv.length ? argv[i + 1] : dflt;
  }
  if (argv.indexOf("--selftest") >= 0) {
    var res = M.selfTest(Number(arg("--seed", 20260910)));
    console.log(JSON.stringify(res, null, 2));
    process.exit(res.ok ? 0 : 1);
  }
  var limit = Number(arg("--limit", 6));
  var status = arg("--status", "all");
  var statuses = status === "all" ? M.DEMO_SEQUENCE : [status];
  M.stream(1, limit, Number(arg("--seed", 20260910)), statuses).forEach(function (f) {
    console.log(JSON.stringify(f));
  });
}

File: backend/mock.py
"""mock.py —— 契约一致的假数据源（《02》任务 B2）。

存在意义（《02》第六节）：**B 线从第一天起就不需要等 A 线。**
只要 A 线还没产出真 JSON，B 线就用本文件在浏览器里看完整仪表盘。

设计要点（这是"好用"和"没用"的差别）：
  - **按 status 反推指标值**，而不是各字段独立乱随机。
    status=fatigue_risk 时 PERCLOS 就必须高、眨眼率就必须低；否则演示时
    "状态写着疲劳、数字却说正常"，评委一眼看穿。
  - 用 random.Random(seed) 局部实例，**不污染全局随机状态**，
    因此同 seed 必得同一序列 → pytest 可以断言，演示可复现。
  - 产出后强制过 contract.validate_frame()，保证与 A 线真实输出同构。

⚠️ 这里的数据是**假的**。报告/PPT 中凡是引用本文件产出的数字，必须标注"Mock 数据"。
"""

from __future__ import annotations

import argparse
import json
import random
import time
from typing import Any, Iterator

try:
    from .contract import STATUS_VALUES, assert_valid, new_frame
except ImportError:
    from contract import STATUS_VALUES, assert_valid, new_frame

# 六态演示时的推荐顺序（B6 验收：六种状态都能演示且样式区分清晰）
DEMO_SEQUENCE = ("normal", "fatigue_risk", "adjust_posture", "unreliable", "disconnected", "done")

_ADVICE = {
    "normal": ("状态正常", "各项指标均在正常范围，信号质量良好"),
    "fatigue_risk": ("疲劳风险升高，建议休息 5 分钟或远眺放松", "触发：PERCLOS 偏高；长闭眼次数增加"),
    "adjust_posture": ("请正对摄像头，保持面部完整出现在画面内", "人脸可见率或头部姿态超出允许范围"),
    "unreliable": ("信号不可靠，暂不输出测量结果", "光照/运动导致信号质量低于门控阈值"),
    "disconnected": ("连接中断，请检查视频源或后端服务", "超过 ws_disconnect_timeout_s 未收到数据帧"),
    "done": ("测量完成", "本次测量正常结束，结果已归档"),
}


def _rng(seed: int, frame_id: int) -> random.Random:
    """每次调用得到独立、可复现的随机源（不用全局 random）。"""
    return random.Random((seed * 1_000_003) ^ (frame_id * 2_654_435_761))


def mock_frame(
    frame_id: int,
    *,
    ts: float | None = None,
    status: str | None = None,
    seed: int = 20260910,
    statuses: tuple[str, ...] = DEMO_SEQUENCE,
    quality_gate: float = 0.75,
) -> dict[str, Any]:
    """生成一帧契约 JSON。status 为 None 时按 statuses 轮转。"""
    if status is None:
        status = statuses[frame_id % len(statuses)]
    if status not in STATUS_VALUES:
        raise ValueError(f"status 必须是 {STATUS_VALUES} 之一，实际 {status!r}")

    r = _rng(seed, frame_id)
    advice, reason = _ADVICE[status]

    # ---- 基线：一切正常 ----
    visible = min(0.999, 0.96 + r.uniform(-0.02, 0.02))
    bbox = (320 + r.randint(-2, 2), 180 + r.randint(-2, 2), 180 + r.randint(-3, 3), 220 + r.randint(-3, 3))
    yaw, pitch, roll = r.uniform(-3, 3), r.uniform(-2, 2), r.uniform(-2, 2)

    ear_l = 0.27 + r.uniform(-0.02, 0.02)
    ear_r = ear_l + r.uniform(-0.01, 0.01)
    blink_state = "OPEN"
    blink_count = 12 + frame_id // 20
    blink_rate = 17.0 + r.uniform(-2, 2)
    perclos = 0.05 + r.uniform(0, 0.03)
    long_close = 0
    mar = 0.10 + r.uniform(0, 0.05)
    yawn = 0
    light = 0.82 + r.uniform(-0.05, 0.05)
    motion = 0.03 + r.uniform(0, 0.03)
    overall = 0.86 + r.uniform(-0.04, 0.04)
    vital: dict[str, Any] = {}

    # ---- 按状态改写指标，保证"状态与数字自洽" ----
    if status == "fatigue_risk":
        perclos = 0.30 + r.uniform(0, 0.12)
        blink_rate = 9.0 + r.uniform(-2, 2)
        long_close = 1 + r.randint(0, 2)
        yawn = 2 + r.randint(0, 2)
        ear_l = 0.20 + r.uniform(-0.02, 0.03)
        ear_r = ear_l + r.uniform(-0.01, 0.01)
        blink_state = "CLOSED" if r.random() < 0.35 else "OPEN"
    elif status == "adjust_posture":
        visible = 0.35 + r.uniform(0, 0.25)
        yaw = 32.0 + r.uniform(0, 14) * (1 if r.random() < 0.5 else -1)
        pitch = r.uniform(-6, 6)
    elif status == "unreliable":
        light = 0.22 + r.uniform(0, 0.22)
        motion = 0.45 + r.uniform(0, 0.4)
        overall = 0.30 + r.uniform(0, 0.25)
        blink_state = "OPEN"
    elif status == "disconnected":
        visible = 0.0
        bbox = (0, 0, 0, 0)
        yaw = pitch = roll = 0.0
        ear_l = ear_r = 0.0
        blink_state = "OPEN"
        perclos = 0.0
        blink_rate = 0.0
        mar = 0.0
        light = motion = overall = 0.0
    elif status == "done":
        perclos = 0.08
        blink_rate = 16.0

    # ---- vital：质量不够就**必须**是 null（产品承诺，与 decision.gate_vitals 一致）----
    if status in ("normal", "fatigue_risk", "done") and overall >= quality_gate:
        hr = 68.0 + r.uniform(-4, 6)
        rr = 15.0 + r.uniform(-2, 2)
        vital = {
            "hr_bpm": round(hr, 1),
            "hr_conf": round(min(0.98, 0.62 + overall * 0.35 + r.uniform(-0.05, 0.05)), 3),
            "rr_per_min": round(rr, 1),
            "rr_conf": round(min(0.95, 0.55 + overall * 0.3 + r.uniform(-0.05, 0.05)), 3),
        }

    frame = new_frame(
        ts=ts if ts is not None else round(time.time(), 3),
        frame_id=frame_id,
        face={
            "visible": round(visible, 3),
            "bbox": [int(v) for v in bbox],
            "pose": {"yaw": round(yaw, 2), "pitch": round(pitch, 2), "roll": round(roll, 2)},
        },
        behavior={
            "ear_left": round(ear_l, 4),
            "ear_right": round(ear_r, 4),
            "blink_state": blink_state,
            "blink_count": blink_count,
            "blink_rate_per_min": round(blink_rate, 2),
            "perclos": round(perclos, 4),
            "long_close_count": long_close,
            "mar": round(mar, 4),
            "yawn_count": yawn,
        },
        vital=vital,
        quality={
            "light_score": round(max(0.0, min(1.0, light)), 4),
            "motion_score": round(max(0.0, min(1.0, motion)), 4),
            "overall": round(max(0.0, min(1.0, overall)), 4),
        },
        status=status,
        advice=advice,
        reason=reason,
    )
    return assert_valid(frame)


def mock_stream(
    hz: float = 1.0,
    *,
    seed: int = 20260910,
    statuses: tuple[str, ...] = DEMO_SEQUENCE,
    realtime: bool = True,
    limit: int | None = None,
    logical_ts: bool = False,
) -> Iterator[dict]:
    """按 hz 持续产出帧。

    realtime=True  → 按 hz 真实 sleep（给前端看的效果）
    realtime=False → 立刻产出（测试用）
    logical_ts=True→ 时间戳用 frame_id/hz 而不是墙上时钟 —— **可复现**，
                     回归测试与截图证据都用这个模式；实时演示用默认的墙上时钟。
    """
    period = 1.0 / hz if hz > 0 else 0.0
    fid = 0
    while limit is None or fid < limit:
        ts = round(fid / hz, 3) if (logical_ts and hz > 0) else None
        yield mock_frame(fid, ts=ts, status=statuses[fid % len(statuses)], seed=seed, statuses=statuses)
        fid += 1
        if realtime and period:
            time.sleep(period)


def main() -> int:
    ap = argparse.ArgumentParser(description="VigiLens 契约一致的 Mock 数据源（B 线用，A 线未就绪时不阻塞）")
    ap.add_argument("--frames", type=int, default=6, help="输出多少帧（默认 6，正好覆盖六态各一帧）")
    ap.add_argument("--hz", type=float, default=1.0, help="实时模式下的推送频率（默认 1 Hz，与契约一致）")
    ap.add_argument("--realtime", action="store_true", help="按 --hz 真实节流（默认不节流，立刻输出）")
    ap.add_argument("--status", default="all", help="all=六态轮转；或指定单个 status 反复输出")
    ap.add_argument("--seed", type=int, default=20260910, help="随机种子（同 seed 必得同序列，便于复现）")
    ap.add_argument("--logical-ts", action="store_true",
                    help="时间戳用 frame_id/hz（可复现，适合归档证据）而不是墙上时钟")
    ap.add_argument("--compact", action="store_true", help="单行紧凑 JSON（默认美化输出）")
    args = ap.parse_args()

    statuses = DEMO_SEQUENCE if args.status == "all" else (args.status,)
    for fr in mock_stream(args.hz, seed=args.seed, statuses=statuses, realtime=args.realtime,
                          limit=args.frames, logical_ts=args.logical_ts):
        if args.compact:
            print(json.dumps(fr, ensure_ascii=False, separators=(",", ":")))
        else:
            print(json.dumps(fr, ensure_ascii=False, indent=2))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
